csrc/host_ir/lower_to_llvm.cpp

```diff
@@ -0,0 +1,912 @@
+template <typename T>
+T ExitOnErr(llvm::Expected<T> &&E) {
+    if (!E) {
+        NVF_ERROR(false, "LLVM JIT Initialization Error: " + llvm::toString(E.takeError()));
+        llvm::errs() << llvm::toString(E.takeError()) << "\n";
+        exit(1);
+    }
+    return std::move(*E);
+}
+
+inline void ExitOnErr(llvm::Error &&Err) {
+  if (Err) {
+    NVF_ERROR(false, "LLVM JIT Initialization Error: " + llvm::toString(std::move(Err)));
+    llvm::errs() << llvm::toString(std::move(Err)) << "\n";
+    exit(1);
+  }
+}
+
+// Helper function to cast iter domains to vals
+std::vector<Val*> domain2vals(const std::vector<IterDomain*>& domain){
+  std::vector<Val*> vals;
+  for(auto* id : domain){
+    vals.push_back(dynamic_cast<Val*>(id));
+  }
+  return vals;
+}
+
+// Helper function to map the current domain to the input domain strictly
+int isSameToInputDomain(std::unordered_map<int, Val*>& boundary_vals, Val* current_domain){
+  for(auto boundary_val : boundary_vals){
+    if(boundary_val.second == current_domain){
+      return boundary_val.first;
+    }
+  }
+  return -1;
+}
+
+// Helper function to check if the current iter domain is alias to the input iter domain
+int mapToInputDomain(std::unordered_map<int, Val*>& boundary_vals, Val* current_domain, const ValGraph& exact_graph){
+  int input_domain_index = isSameToInputDomain(boundary_vals, current_domain);
+  if(input_domain_index != -1){
+    return input_domain_index;
+  }
+  for(auto boundary_val : boundary_vals){ 
+    if(exact_graph.disjointValSets().strictAreMapped(boundary_val.second, current_domain)){
+      return boundary_val.first;
+    }
+  } 
+  return -1;
+}
+
+/*
+
+Generate LLVM IR for a dependency graph
+By default, we assume it is in typological order, which means input values are ready to use
+
+*/
+void generate_shape_llvm_ir(Expr* expr, llvm::IRBuilder<>& builder, std::unordered_map<ValGroup,llvm::Value*>& val2llvm, std::unordered_map<int, Val*>& boundary_vals, const ValGraph& graph) {
+  std::string op_string = std::string(expr->getOpString());
+
+  // Perform the merge -> mul transformation
+  if(op_string == "Merge"){
+    llvm::Value* result = nullptr;
+    auto* merge_expr = expr->as<Merge>();
+    auto* merge_input_outer_val = merge_expr->outer()->as<Val>();
+    auto* merge_input_inner_val = merge_expr->inner()->as<Val>();
+    auto* merge_output_val = merge_expr->outputs()[0]->as<Val>();
+
+    int input_outer_potential_index = mapToInputDomain(boundary_vals, merge_input_outer_val, graph);
+    int input_inner_potential_index = mapToInputDomain(boundary_vals, merge_input_inner_val, graph);
+    llvm::Value* input_outer_llvm_val = nullptr;
+    llvm::Value* input_inner_llvm_val = nullptr;
+
+    if(input_outer_potential_index != -1){
+      input_outer_llvm_val = val2llvm[graph.toGroup(boundary_vals[input_outer_potential_index])];
+    }
+    else{
+      input_outer_llvm_val = val2llvm[graph.toGroup(merge_input_outer_val)];
+    }
```

Comment:
I'm seeing this pattern a lot vvv
```
auto* v = ...;
int index = mapToInputDomain(...);
llvm::Value* val = nullptr;
if (index != -1) {
  val = val2llvm[graph.toGroup(boundary_vals[index])];
} else {
  val = val2llvm[graph.toGroup(v)];
}
```

Maybe make it a lambda to keep the code easier to read.